torch/_subclasses/fake_tensor.py

```diff
@@ -1414,6 +1524,215 @@ def __exit__(self, a, b, c):
+            len(FakeTensorMode.cache),
+        )
+
+    @classmethod
+    def cache_clear(cls):
+        """
+        Clear the dispatch cache.
+        """
+        cls.hits = 0
+        cls.misses = 0
+        cls.cache_bypasses.clear()
+        cls.cache.clear()
+
+    def _cached_dispatch_impl(
+        self,
+        func: OpOverload,
+        types: Tuple[Any],
+        args: Tuple[Any],
+        kwargs: Dict[str, Any],
+    ):
+        """
+        Lookup a cache entry for the given arguments. If none exists, dispatch
+        and cache the result (if the result is cache-able).
+        """
+        try:
+            key = self._cache_key(func, args, kwargs)
+        except _BypassDispatchCache as e:
+            FakeTensorMode.cache_bypasses[e.reason] += 1
+            return self._dispatch_impl(func, types, args, kwargs)
+
+        entry = FakeTensorMode.cache.get(key, None)
+        if entry is not None:
+            FakeTensorMode.cache_hits += 1
+            output = self._output_from_cache_entry(entry, func, args)
+        else:
+            FakeTensorMode.cache_misses += 1
+            output = self._dispatch_impl(func, types, args, kwargs)
+            try:
+                self._create_cache_entry(key, args, output)
+            except _BypassDispatchCache as e:
+                FakeTensorMode.cache_bypasses[e.reason] += 1
+
+        return output
+
+    def _cache_key(
+        self, func: OpOverload, args: Tuple[Any], kwargs: Dict[str, Any]
+    ) -> _DispatchCacheKey:
+        """
+        Create a cache key given the dispatch args. Raise _BypassDispatchCache
+        for any situation that precludes caching.
+        """
+        # Avoid caching for any ops that would require a more sophisticated
+        # caching implementation, e.g., data dependent ops or ops that modify
+        # the inputs.
+        if torch.Tag.data_dependent_output in func.tags:
+            raise _BypassDispatchCache("data dependent output")
+
+        if torch.Tag.dynamic_output_shape in func.tags:
+            raise _BypassDispatchCache("dynamic output shape")
+
+        if torch.Tag.inplace_view in func.tags:
+            raise _BypassDispatchCache("inplace view")
+
+        if func == aten._unsafe_view.default:
+            raise _BypassDispatchCache("unsafe view")
+
+        if func == aten.set_.source_Tensor:
+            raise _BypassDispatchCache("set source")
+
+        if func in self.lift_fns:
+            raise _BypassDispatchCache("lift")
+
+        # In order to handle storage aliasing, we need to establish the alias
+        # for any view op on a cache hit. But CompositeImplicitAutograd ops may
+        # or may not alias the input, so just punt on caching these.
+        if func.is_view and torch._C._dispatch_has_kernel_for_dispatch_key(
+            func.name(), torch._C.DispatchKey.CompositeImplicitAutograd
+        ):
+            raise _BypassDispatchCache("CompositeImplicitAutograd")
```

Comment:
cc @bdhirsh, you were talking about an easier way of doing this right ?

Reply:
the code above is the "easier way" I had in mind 😛 .

This code is basically querying the source of truth (the dispatcher state), to ask if there are any CompositeImplicitAutograd registrations

Reply:
In theory we could cache on some CompositeImplicitAutograd ops, the problem is that we haven't actually done the work to find and test for the set of CompositeImplicitAutograd ops that are not cache friendly. So instead we can just decompose them all here (aka what we're doing right now)